clients/drcachesim/tools/trace_filter.cpp

```diff
@@ -0,0 +1,217 @@
+        !directory_iterator_t::is_directory(output_dir)) {
+        ERRMSG("Trace dir or output dir is not a directory\n");
+        return false;
+    }
+    directory_iterator_t end;
+    directory_iterator_t iter(trace_dir);
+    if (!iter) {
+        ERRMSG("Failed to list directory %s: %s", trace_dir.c_str(),
+               iter.error_string().c_str());
+        return false;
+    }
+    for (; iter != end; ++iter) {
+        const std::string fname = *iter;
+        if (fname == "." || fname == "..")
+            continue;
+        const std::string trace_path = trace_dir + DIRSEP + fname;
+        const std::string output_path = output_dir + DIRSEP + fname;
+        std::unique_ptr<trace_entry_reader_t> reader = get_reader(trace_path, verbosity);
+        if (!reader) {
+            error_string_ = "Could not get a reader for " + trace_path;
+            return false;
+        }
+
+        std::unique_ptr<std::ostream> writer = get_writer(output_path, verbosity);
+        if (!writer) {
+            error_string_ = "Could not get a writer for " + output_path;
+            return false;
+        }
+
+        thread_data_.push_back(shard_data_t(static_cast<int>(thread_data_.size()),
+                                            std::move(reader), std::move(writer),
+                                            trace_path, output_path));
+    }
+    // Like raw2trace, we use a simple round-robin static work assigment.  This
+    // could be improved later with dynamic work queue for better load balancing.
+    if (worker_count_ <= 0)
+        worker_count_ = std::thread::hardware_concurrency();
+    worker_tasks_.resize(worker_count_);
+    int worker = 0;
+    for (size_t i = 0; i < thread_data_.size(); ++i) {
+        VPRINT(this, 2, "Worker %d assigned trace shard %zd\n", worker, i);
+        worker_tasks_[worker].push_back(&thread_data_[i]);
+        thread_data_[i].worker = worker;
+        worker = (worker + 1) % worker_count_;
+    }
+    trace_end_ = std::unique_ptr<default_trace_entry_file_reader_t>(
+        new default_trace_entry_file_reader_t());
+    return true;
+}
+
+void
+trace_filter_t::process_tasks(std::vector<shard_data_t *> *tasks)
+{
+    if (tasks->empty()) {
+        VPRINT(this, 1, "Worker has no tasks\n");
+        return;
+    }
+    VPRINT(this, 1, "Worker %d assigned %zd task(s)\n", (*tasks)[0]->worker,
+           tasks->size());
+    for (shard_data_t *tdata : *tasks) {
+        VPRINT(this, 1, "Worker %d starting on trace shard %d\n", tdata->worker,
+               tdata->index);
+        if (!tdata->iter->init()) {
+            tdata->error = "Failed to read from trace" + tdata->trace_file;
+            return;
+        }
+        for (; *tdata->iter != *trace_end_; ++(*tdata->iter)) {
+            const trace_entry_t &entry = **tdata->iter;
+            // TODO i#5675: Filter entries here.
+            if (!tdata->writer->write((char *)&entry, sizeof(entry))) {
```

Comment:
What if we want to output multiple files?  E.g., we identify 5 regions of interest and want to create separate files for each (as opposed to just inserting markers).  Maybe that's harder to support if we allow multiple filter tools where just one output makes sense so maybe we do not support that but it is worth acknowledging as an explicit design decision.

Reply:
Added a comment.